File: java/linkedlist/single/Node.java
package com.company;

/**
 * Created by junkyu on 2017. 3. 22..
 */

/*
* Single Linked List Node
* */
public class Node<T> {
    private T value;
    private Node<T> next;

    public Node(){
    }

    public Node(T value){
        this.value = value;
        this.next = null;
    }

    public T getValue(){
       return this.value;
    }

    public void setValue(T value){
        this.value = value;
    }

    public Node getNext(){
        return this.next;
    }

    public void setNext(Node node){
        this.next = node;
    }

    public void appendToTail(T value){
       Node tail = new Node(value);
       Node n = this;

       while(n.next != null){
           n = n.next;
       }
       n.setNext(tail);
    }

    public Node deleteNode(Node head, T value){
        Node node = head;

        if(head.value == value){
            return head.next;
        }

        while(node.next != null){
            if(node.next.value == value){
                node.next = node.next.next;
                return head;
            }
            node = node.next;
        }

        return head;
    }

    public static Node nthToLastNonRecursive(Node head, int index){
        /*Non Recursive*/
        Node curNode = head;
        int length = 0;
        int valueIndex = 0;

        while(curNode != null){
            length++;
            curNode = curNode.getNext();
        }

        curNode = head;
        valueIndex = length - 1 - index;
        for(int i = 0 ; i < valueIndex; i++ ){
            curNode = curNode.getNext();
        }

        return curNode;

    }

    public static Node nthToLastRecursive(Node head, int index, IntegerWrapper indexVar){
        /* Recursive */
        if(head.getNext() == null){
            return null;
        }
        Node node = nthToLastRecursive(head.getNext(), index, indexVar);
        indexVar.setCount(indexVar.getCount() + 1);
        if(indexVar.getCount() == index){
            return head;
        }

        return node;
    }
}
class IntegerWrapper{
    private int count;

    public IntegerWrapper(){
        this.count = 0;
    }
    public void setCount(int count){
        this.count = count;
    }
    public int getCount(){
        return this.count;
    }
}

File: python/linked_list/node.py
class Node():
    """ Single Linked List Node """
    def __init__(self, value=None, next=None):
        self.__value = value
        self.__next = next

    @property
    def value(self):
        return self.__value

    @value.setter
    def value(self, value):
        self.__value = value

    @property
    def next(self):
        return self.__next

    @next.setter
    def next(self, next):
        self.__next = next
    
    def append_to_tail(self, node):
        curr_node = self

        while curr_node.next != None:
            curr_node = curr_node.next
        
        curr_node.next = node
    
    @classmethod
    def delete_node(self, head, value):
        if head.value == value:
            head = head.next
            return

        curr = head

        while curr.next != None:
            if curr.next.value == value:
                curr.next = curr.next.next
                return
            
            curr = curr.next
    
    def track_list(self):
        curr = self

        print("HEAD", end=' -> ')
        while curr != None:
            print("%r" % curr.value, end=' -> ')
            curr = curr.next

        print("None")
